Implement va_list and va_arg for s390x FFI

uweigand opened the pull request:
Following the s390x ELF ABI and based on the clang implementation, provide appropriate definitions of va_list in library/core/src/ffi/mod.rs and va_arg handling in compiler/rustc_codegen_llvm/src/va_arg.rs.

Fixes the following test cases on s390x:
src/test/run-make-fulldeps/c-link-to-rust-va-list-fn src/test/ui/abi/variadic-ffi.rs

Fixes https://github.com/rust-lang/rust/issues/84628.

uweigand committed:
Implement va_list and va_arg for s390x FFI

Following the s390x ELF ABI and based on the clang implementation,
provide appropriate definitions of va_list in library/core/src/ffi/mod.rs
and va_arg handling in compiler/rustc_codegen_llvm/src/va_arg.rs.

Fixes the following test cases on s390x:
src/test/run-make-fulldeps/c-link-to-rust-va-list-fn
src/test/ui/abi/variadic-ffi.rs

Fixes https://github.com/rust-lang/rust/issues/84628.

cuviper commented:
@davidtwco if you're not comfortable looking at s390x, I can take review on this.

dlrobertson reviewed (commented):
This is awesome! Thanks for the work! I don't know a ton about s390x, but is va_end by chance a nop? See [this comment](https://github.com/rust-lang/rust/blob/7210e46dc69a4b197a313d093fe145722c248b7d/library/core/src/ffi/mod.rs#L554-L567).

uweigand commented:
> This is awesome! Thanks for the work! I don't know a ton about s390x, but is va_end by chance a nop? See [this comment](https://github.com/rust-lang/rust/blob/7210e46dc69a4b197a313d093fe145722c248b7d/library/core/src/ffi/mod.rs#L554-L567).

Yes, `va_end` is a no-op on s390x as well.

dlrobertson commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -175,6 +175,92 @@ fn emit_aapcs_va_arg<'ll, 'tcx>(
+    // FIXME: structs with a single element of floating-point type should be
+    // passed in a FPR instead of a GPR.
+    let gpr_type = indirect || target_ty.is_any_ptr() || target_ty.is_integral();
+    let (max_regs, reg_count, reg_save_index, reg_padding) = if gpr_type {
```
nit: I like the way clang does this with a `reg_count_field` here (either 0 or 1) followed by something like:

```
let reg_count = bx.struct_gep(...,reg_count_field))
```

uweigand replied:
Done.

dlrobertson commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -175,6 +175,92 @@ fn emit_aapcs_va_arg<'ll, 'tcx>(
+    let reg_ptr =
+        bx.struct_gep(va_list_ty, va_list_addr, va_list_layout.llvm_field_index(bx.cx, 3));
+    let reg_ptr_v = bx.load(bx.type_i8p(), reg_ptr, bx.tcx().data_layout.pointer_align.abi);
+    let reg_off = bx.mul(reg_count_v, bx.const_u64(8));
```
nit: could we use something more clang-like (e.g. `scaled_reg_count`). I have no real strong opinions, but it would be nice to avoid re-using `reg_off`.

uweigand replied:
Done.

dlrobertson commented on compiler/rustc_codegen_llvm/src/va_arg.rs:
```diff
@@ -175,6 +175,92 @@ fn emit_aapcs_va_arg<'ll, 'tcx>(
+    let target_ty_size = bx.cx.size_of(target_ty).bytes();
+    let indirect: bool = target_ty_size > 8 || !target_ty_size.is_power_of_two();
+    let unpadded_size = if indirect { 8 } else { target_ty_size };
+    let padded_size = 8;
```
`padded_size - unpadded_size` is used at least twice, would it make sense to add a variable for this?

uweigand replied:
Done.

dlrobertson reviewed (commented):
Thanks again for working on this! Looks reasonable after look at clangs implementation followed by a review of this, but note that I have no experience with s390x. Also not that I don't have a box to test this on.

uweigand commented:
> Thanks again for working on this! Looks reasonable after look at clangs implementation followed by a review of this, but note that I have no experience with s390x. Also not that I don't have a box to test this on.

Thanks for the review!  I've addressed your comments, and also resolved the one FIXME about aggregate handling.

FWIW I'm pretty confident that the s390x ABI parts are correct (I'm the code owner of the IBM Z LLVM back-end, and one of the authors of the platform ABI specification); if there's still a bug somewhere, it's more likely to be about the Rust compiler implementation details, which I'm not very familar with.

I have run the full test suite (`./x.py test`) natively on a s390x machine, and the patch does fix the two previously failing test cases mentioned above.  If there are any further tests you'd like me to run, please let me know.

dlrobertson reviewed (approved):
> if there's still a bug somewhere, it's more likely to be about the Rust compiler implementation details, which I'm not very familar with.

Looks like you got those bits right!

> FWIW I'm pretty confident that the s390x ABI parts are correct (I'm the code owner of the IBM Z LLVM back-end, and one of the authors of the platform ABI specification);

👍 just wanted to spot check, and learn a bit about s390x 😃

dlrobertson commented:
I'm unable to r+ this, but this looks great!

CC @joshtriplett @cuviper

nikic commented:
For reference: https://github.com/llvm/llvm-project/blob/e8bcc37fff5bda7dd9326903a2c31e6703b4fe68/clang/lib/CodeGen/TargetInfo.cpp#L7602

nikic commented:
@bors r+

bors commented:
:pushpin: Commit eb22d70aed39614dd9318e3fa1bf91846d74b8aa has been approved by `nikic`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikic eb22d70aed39614dd9318e3fa1bf91846d74b8aa -->
<!-- homu: {"type":"Approved","sha":"eb22d70aed39614dd9318e3fa1bf91846d74b8aa","approver":"nikic","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit eb22d70aed39614dd9318e3fa1bf91846d74b8aa with merge d6da428f343ab811b2b132364360ba13ff05830c...
<!-- homu: {"type":"BuildStarted","head_sha":"eb22d70aed39614dd9318e3fa1bf91846d74b8aa","merge_sha":"d6da428f343ab811b2b132364360ba13ff05830c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3742625756/jobs/6356146390)
Approved by: nikic
Pushing d6da428f343ab811b2b132364360ba13ff05830c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikic","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3742625756/jobs/6356146390"},"merge_sha":"d6da428f343ab811b2b132364360ba13ff05830c"} -->

rust-timer commented:
Finished benchmarking commit (d6da428f343ab811b2b132364360ba13ff05830c): [comparison URL](https://perf.rust-lang.org/compare.html?start=65bd2a6a73d6a74fb1266a1d96b23de8810a5fb2&end=d6da428f343ab811b2b132364360ba13ff05830c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=65bd2a6a73d6a74fb1266a1d96b23de8810a5fb2&end=d6da428f343ab811b2b132364360ba13ff05830c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.2%, -1.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.6%, -2.0%] | 6     |
| All ❌✅ (primary)                 | -1.1% | [-1.2%, -1.0%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=65bd2a6a73d6a74fb1266a1d96b23de8810a5fb2&end=d6da428f343ab811b2b132364360ba13ff05830c&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=65bd2a6a73d6a74fb1266a1d96b23de8810a5fb2&end=d6da428f343ab811b2b132364360ba13ff05830c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

dlrobertson commented:
\o/ Thanks @uweigand for working on this!
